lib/jxl/image.h

```diff
@@ -194,21 +194,65 @@ class Image3;
 // shifting the pointer by x0/y0 allows this to apply to multiple images with
 // different resolutions (e.g. color transform and quantization field).
 // Can compare using SameSize(rect1, rect2).
-class Rect {
+template <typename T>
+class RectT {
  public:
   // Most windows are xsize_max * ysize_max, except those on the borders where
   // begin + size_max > end.
-  constexpr Rect(size_t xbegin, size_t ybegin, size_t xsize_max,
-                 size_t ysize_max, size_t xend, size_t yend)
+  constexpr RectT(T xbegin, T ybegin, size_t xsize_max, size_t ysize_max,
+                  T xend, T yend)
       : x0_(xbegin),
         y0_(ybegin),
         xsize_(ClampedSize(xbegin, xsize_max, xend)),
         ysize_(ClampedSize(ybegin, ysize_max, yend)) {}
 
   // Construct with origin and known size (typically from another Rect).
-  constexpr Rect(size_t xbegin, size_t ybegin, size_t xsize, size_t ysize)
+  constexpr RectT(T xbegin, T ybegin, size_t xsize, size_t ysize)
       : x0_(xbegin), y0_(ybegin), xsize_(xsize), ysize_(ysize) {}
 
+  RectT(const RectT&) = default;
+  RectT& operator=(const RectT&) = default;
+
+  JXL_MUST_USE_RESULT RectT Intersection(const RectT& other) const {
+    return RectT(std::max(x0_, other.x0_), std::max(y0_, other.y0_), xsize_,
+                 ysize_, std::min(x0_ + xsize_, other.x0_ + other.xsize_),
+                 std::min(y0_ + ysize_, other.y0_ + other.ysize_));
+  }
+
+  JXL_MUST_USE_RESULT RectT Translate(int64_t x_offset,
+                                      int64_t y_offset) const {
+    return RectT(x0_ + x_offset, y0_ + y_offset, xsize_, ysize_);
+  }
+
+  T x0() const { return x0_; }
+  T y0() const { return y0_; }
+  size_t xsize() const { return xsize_; }
+  size_t ysize() const { return ysize_; }
+
+ protected:
+  // Returns size_max, or whatever is left in [begin, end).
+  static constexpr size_t ClampedSize(T begin, size_t size_max, T end) {
+    return (static_cast<ssize_t>(begin + size_max) <= static_cast<ssize_t>(end))
+               ? size_max
+               : (end > begin ? end - begin : 0);
+  }
+
+  T x0_;
+  T y0_;
+
+  size_t xsize_;
+  size_t ysize_;
+};
+
+class Rect : public RectT<size_t> {
+ public:
+  constexpr Rect(size_t xbegin, size_t ybegin, size_t xsize_max,
```

Comment:
Could do using RectT::RectT, no?

Reply:
Merged back Rect and RectT. Inheritance of templatized producer without CRTP is a hell, and with it - it looks disgusting.

Reply:
Just thinking out loud, but couldn't we make all rects use ssize_t?

Reply:
We could try. One thing I foresee is problem with large images on 32-bit system... But it is likely, we will get OOM first.